



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
        <link rel="canonical" href="http://dib-lab.github.io/eelpond/DE/">
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching pages">
      
        <meta name="lang:search.result.one" content="1 matching page">
      
        <meta name="lang:search.result.other" content="# matching page">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="..">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-3.0.3">
    
    
      
        <title>Differential Expression - eelpond</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.451f80e5.css">
      
        <link rel="stylesheet" href="../assets/stylesheets/application-palette.22915126.css">
      
      
        
        
        <meta name="theme-color" content="#ab47bc">
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../css/custom.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="purple" data-md-color-accent="purple">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="../#differential-expression-analysis-with-deseq2" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="http://dib-lab.github.io/eelpond/" title="eelpond" class="md-header-nav__button md-logo">
          
            <i class="md-icon">dns</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                eelpond
              </span>
              <span class="md-header-nav__topic">
                Differential Expression
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search this site" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching this site
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/dib-lab/eelpond" title="Go to repository" class="md-source" data-md-source="github">
    
    <div class="md-source__repository">
      eelpond
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="http://dib-lab.github.io/eelpond/" title="eelpond" class="md-nav__button md-logo">
      
        <i class="md-icon">dns</i>
      
    </a>
    eelpond
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/dib-lab/eelpond" title="Go to repository" class="md-source" data-md-source="github">
    
    <div class="md-source__repository">
      eelpond
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="About" class="md-nav__link">
      About
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../QC/" title="Read Quality Trimming and Filtering" class="md-nav__link">
      Read Quality Trimming and Filtering
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Annotation/" title="Annotation" class="md-nav__link">
      Annotation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Diginorm/" title="Digital Normalization" class="md-nav__link">
      Digital Normalization
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Assembly/" title="Assembly" class="md-nav__link">
      Assembly
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Quality/" title="Quality Assessment" class="md-nav__link">
      Quality Assessment
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Quant/" title="Transcript Quantification" class="md-nav__link">
      Transcript Quantification
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Differential Expression
      </label>
    
    <a href="./" title="Differential Expression" class="md-nav__link md-nav__link--active">
      Differential Expression
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#rstudio" title="RStudio!" class="md-nav__link">
    RStudio!
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-9" type="checkbox" id="nav-9">
    
    <label class="md-nav__link" for="nav-9">
      For Developers
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-9">
        For Developers
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../dev_tips/" title="Notes for Developers" class="md-nav__link">
      Notes for Developers
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#rstudio" title="RStudio!" class="md-nav__link">
    RStudio!
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="differential-expression-analysis-with-deseq2">Differential expression analysis with DESeq2<a class="headerlink" href="#differential-expression-analysis-with-deseq2" title="Permanent link">&para;</a></h1>
<p>Comparing gene expression differences in samples between experimental conditions. </p>
<p>We will be using <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>.</p>
<p>References:
<em> <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">Documentation for DESeq2 with example analysis</a>
</em> <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">Love et al. 2014</a>
* <a href="https://www.nature.com/nbt/journal/v34/n12/full/nbt.3682.html">Love et al. 2016</a></p>
<p>Additional links:
<em> <a href="https://github.com/Open-Data-Science-at-SIO/RNAseq-workshop-2017/blob/master/_static/Jane_differential_expression.pdf">DE lecture by Jane Khudyakov, July 2017</a>
</em> <a href="http://htmlpreview.github.io/?https://github.com/ljcohen/Fhet_MDPL_MDPP_salinity_DE/blob/master/Fhet_MDPL_v_MDPP_interactiononly_FW_BW.html">Example DE analysis from two populations of killifish! (Fundulus heteroclitus MDPL vs. MDPL)</a>
* <a href="https://github.com/ljcohen/ECE221_final_project/blob/master/Cohen_Li_ECE221_review-differential-gene.pdf">A Review of Differential Gene Expression Software for mRNA sequencing</a></p>
<h2 id="rstudio">RStudio!<a class="headerlink" href="#rstudio" title="Permanent link">&para;</a></h2>
<p>The pipeline will be running these commands in an R script. You could run them in R Studio:</p>
<p>Load libraries</p>
<div class="codehilite"><pre><span></span>library(DESeq2)
library(&quot;lattice&quot;)
library(tximport)
library(readr)
library(gplots)
library(RColorBrewer)
source(&#39;~/plotPCAWithSampleNames.R&#39;)
</pre></div>


<p>Tell RStudio where your files are and ask whether they exist:</p>
<div class="codehilite"><pre><span></span>setwd(&quot;/mnt/work/quant/salmon_out/&quot;)
dir&lt;-&quot;/mnt/work/quant/&quot;
files_list = list.files()
files &lt;- file.path(dir, &quot;salmon_out&quot;,files_list, &quot;quant.sf&quot;)
names(files) &lt;- c(&quot;0Hour_1&quot;,&quot;0Hour_2&quot;,&quot;0Hour_3&quot;,&quot;0Hour_4&quot;,&quot;0Hour_5&quot;,&quot;6Hour_1&quot;,&quot;6Hour_2&quot;,&quot;6Hour_3&quot;,&quot;6Hour_4&quot;,&quot;6Hour_5&quot;)
files
print(file.exists(files))
</pre></div>


<p>Grab the <a href="https://raw.githubusercontent.com/Open-Data-Science-at-SIO/RNAseq-workshop-2017/master/_static/nema_transcript_gene_id.txt">gene names</a> and transcript ID file to <a href="https://f1000research.com/articles/4-1521/v2">summarize expression at the gene level</a>.</p>
<div class="codehilite"><pre><span></span>tx2gene &lt;- read.table(&quot;~/nema_transcript_gene_id.txt&quot;,sep=&quot;\t&quot;)
cols&lt;-c(&quot;transcript_id&quot;,&quot;gene_id&quot;)
colnames(tx2gene)&lt;-cols
head(tx2gene)
txi.salmon &lt;- tximport(files, type = &quot;salmon&quot;, tx2gene = tx2gene,importer=read.delim)
head(txi.salmon$counts)
dim(txi.salmon$counts)
</pre></div>


<p>Assign experimental variables:</p>
<div class="codehilite"><pre><span></span>condition = factor(c(&quot;0Hour&quot;,&quot;0Hour&quot;,&quot;0Hour&quot;,&quot;0Hour&quot;,&quot;0Hour&quot;,&quot;6Hour&quot;,&quot;6Hour&quot;,&quot;6Hour&quot;,&quot;6Hour&quot;,&quot;6Hour&quot;))
ExpDesign &lt;- data.frame(row.names=colnames(txi.salmon$counts), condition = condition)
ExpDesign
</pre></div>


<p>Run DESeq2:</p>
<div class="codehilite"><pre><span></span>dds &lt;- DESeqDataSetFromTximport(txi.salmon, ExpDesign, ~condition)
dds &lt;- DESeq(dds, betaPrior=FALSE)
</pre></div>


<p>Get counts:</p>
<div class="codehilite"><pre><span></span>counts_table = counts( dds, normalized=TRUE )
</pre></div>


<p>Filtering out low expression transcripts:</p>
<p>See plot from <a href="https://github.com/Open-Data-Science-at-SIO/RNAseq-workshop-2017/blob/master/_static/Before-after_filter.pdf">Lisa Komoroske</a> generated with <a href="https://www.bioconductor.org/help/workflows/RNAseq123/">RNAseq123</a></p>
<div class="codehilite"><pre><span></span>filtered_norm_counts&lt;-counts_table[!rowSums(counts_table==0)&gt;=1, ]
filtered_norm_counts&lt;-as.data.frame(filtered_norm_counts)
GeneID&lt;-rownames(filtered_norm_counts)
filtered_norm_counts&lt;-cbind(filtered_norm_counts,GeneID)
dim(filtered_norm_counts)
head(filtered_norm_counts)
</pre></div>


<p>Estimate dispersion:</p>
<div class="codehilite"><pre><span></span>plotDispEsts(dds)
</pre></div>


<p>PCA:</p>
<div class="codehilite"><pre><span></span>log_dds&lt;-rlog(dds)
plotPCAWithSampleNames(log_dds, intgroup=&quot;condition&quot;, ntop=40000)
</pre></div>


<p>Get DE results:</p>
<div class="codehilite"><pre><span></span>res&lt;-results(dds,contrast=c(&quot;condition&quot;,&quot;6Hour&quot;,&quot;0Hour&quot;))
head(res)
res_ordered&lt;-res[order(res$padj),]
GeneID&lt;-rownames(res_ordered)
res_ordered&lt;-as.data.frame(res_ordered)
res_genes&lt;-cbind(res_ordered,GeneID)
dim(res_genes)
head(res_genes)
dim(res_genes)
res_genes_merged &lt;- merge(res_genes,filtered_norm_counts,by=unique(&quot;GeneID&quot;))
dim(res_genes_merged)
head(res_genes_merged)
res_ordered&lt;-res_genes_merged[order(res_genes_merged$padj),]
write.csv(res_ordered, file=&quot;nema_DESeq_all.csv&quot; )
</pre></div>


<p>Set a threshold cutoff of padj&lt;0.05 and ± log2FC 1:</p>
<div class="codehilite"><pre><span></span>resSig = res_ordered[res_ordered$padj &lt; 0.05, ]
resSig = resSig[resSig$log2FoldChange &gt; 1 | resSig$log2FoldChange &lt; -1,]
write.csv(resSig,file=&quot;nema_DESeq_padj0.05_log2FC1.csv&quot;)
</pre></div>


<p>MA plot with gene names:</p>
<div class="codehilite"><pre><span></span>plot(log2(res_ordered$baseMean), res_ordered$log2FoldChange, col=ifelse(res_ordered$padj &lt; 0.05, &quot;red&quot;,&quot;gray67&quot;),main=&quot;nema (padj&lt;0.05, log2FC = ±1)&quot;,xlim=c(1,20),pch=20,cex=1,ylim=c(-12,12))
abline(h=c(-1,1), col=&quot;blue&quot;)
genes&lt;-resSig$GeneID
mygenes &lt;- resSig[,]
baseMean_mygenes &lt;- mygenes[,&quot;baseMean&quot;]
log2FoldChange_mygenes &lt;- mygenes[,&quot;log2FoldChange&quot;]
text(log2(baseMean_mygenes),log2FoldChange_mygenes,labels=genes,pos=2,cex=0.60)
</pre></div>


<p>Heatmap</p>
<div class="codehilite"><pre><span></span>d&lt;-resSig
dim(d)
head(d)
colnames(d)
d&lt;-d[,c(8:17)]
d&lt;-as.matrix(d)
d&lt;-as.data.frame(d)
d&lt;-as.matrix(d)
rownames(d) &lt;- resSig[,1]
head(d)

hr &lt;- hclust(as.dist(1-cor(t(d), method=&quot;pearson&quot;)), method=&quot;complete&quot;)
mycl &lt;- cutree(hr, h=max(hr$height/1.5))
clusterCols &lt;- rainbow(length(unique(mycl)))
myClusterSideBar &lt;- clusterCols[mycl]
myheatcol &lt;- greenred(75)
heatmap.2(d, main=&quot;nema (padj&lt;0.05, log2FC = ±1)&quot;, 
          Rowv=as.dendrogram(hr),
          cexRow=0.75,cexCol=0.8,srtCol= 90,
          adjCol = c(NA,0),offsetCol=2.5, 
          Colv=NA, dendrogram=&quot;row&quot;, 
          scale=&quot;row&quot;, col=myheatcol, 
          density.info=&quot;none&quot;, 
          trace=&quot;none&quot;, RowSideColors= myClusterSideBar)
</pre></div>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../Quant/" title="Transcript Quantification" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Transcript Quantification
              </span>
            </div>
          </a>
        
        
          <a href="../dev_tips/" title="Notes for Developers" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Notes for Developers
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2018 <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a> at UC Davis
          </div>
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.e72fd936.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
        <script src="../search/main.js"></script>
      
    
    
      
    
  </body>
</html>