
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <link rel="canonical" href="http://dib-lab.github.io/elvers/dammit/">
      
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.0.1">
    
    
      
        <title>dammit - elvers</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.38780c08.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.3f72e892.min.css">
        
          
          
          <meta name="theme-color" content="">
        
      
    
    
    
      
        
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="darkblue" data-md-color-accent="blue">
      
  
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#annotating-de-novo-transcriptomes-with-dammit" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="http://dib-lab.github.io/elvers" title="elvers" class="md-header-nav__button md-logo" aria-label="elvers">
      
  <img src="../img/eel.svg" alt="logo">

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      
        <div class="md-header-nav__ellipsis">
          <span class="md-header-nav__topic md-ellipsis">
            elvers
          </span>
          <span class="md-header-nav__topic md-ellipsis">
            
              dammit
            
          </span>
        </div>
      
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active">
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/dib-lab/elvers" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    elvers
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="http://dib-lab.github.io/elvers" title="elvers" class="md-nav__button md-logo" aria-label="elvers">
      
  <img src="../img/eel.svg" alt="logo">

    </a>
    elvers
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/dib-lab/elvers" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    elvers
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="elvers" class="md-nav__link">
      elvers
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../about/" title="About" class="md-nav__link">
      About
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../quickstart/" title="Quickstart" class="md-nav__link">
      Quickstart
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../configure/" title="Choosing and Configuring Workflows" class="md-nav__link">
      Choosing and Configuring Workflows
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5" checked>
    
    <label class="md-nav__link" for="nav-5">
      Available Workflows
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Available Workflows" data-md-level="1">
      <label class="md-nav__title" for="nav-5">
        <span class="md-nav__icon md-icon"></span>
        Available Workflows
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../eel_pond_workflow/" title="Eel Pond Protocol" class="md-nav__link">
      Eel Pond Protocol
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../protein_assembly_workflow/" title="Protein-level Assembly Workflow" class="md-nav__link">
      Protein-level Assembly Workflow
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5-3" type="checkbox" id="nav-5-3">
    
    <label class="md-nav__link" for="nav-5-3">
      Subworkflows
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Subworkflows" data-md-level="2">
      <label class="md-nav__title" for="nav-5-3">
        <span class="md-nav__icon md-icon"></span>
        Subworkflows
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../preprocess/" title="preprocess" class="md-nav__link">
      preprocess
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../kmer_trim/" title="kmer_trim" class="md-nav__link">
      kmer_trim
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../assemble/" title="assemble" class="md-nav__link">
      assemble
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../annotate/" title="annotate" class="md-nav__link">
      annotate
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../quantify/" title="quantify" class="md-nav__link">
      quantify
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../diffexp/" title="diffexp" class="md-nav__link">
      diffexp
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../sourmash_compute/" title="sourmash_compute" class="md-nav__link">
      sourmash_compute
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../plass_assemble/" title="plass_assemble" class="md-nav__link">
      plass_assemble
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../paladin_map/" title="paladin_map" class="md-nav__link">
      paladin_map
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-5-4" type="checkbox" id="nav-5-4" checked>
    
    <label class="md-nav__link" for="nav-5-4">
      Programs Used
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="Programs Used" data-md-level="2">
      <label class="md-nav__title" for="nav-5-4">
        <span class="md-nav__icon md-icon"></span>
        Programs Used
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../get_data/" title="get_data" class="md-nav__link">
      get_data
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../fastqc/" title="fastqc" class="md-nav__link">
      fastqc
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../trimmomatic/" title="trimmomatic" class="md-nav__link">
      trimmomatic
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../khmer/" title="khmer" class="md-nav__link">
      khmer
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../trinity/" title="trinity" class="md-nav__link">
      trinity
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../get_reference/" title="get_reference" class="md-nav__link">
      get_reference
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        dammit
        <span class="md-nav__icon md-icon"></span>
      </label>
    
    <a href="./" title="dammit" class="md-nav__link md-nav__link--active">
      dammit
    </a>
    
      
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#computational-requirements" class="md-nav__link">
    Computational Requirements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-commands" class="md-nav__link">
    Dammit Commands
  </a>
  
    <nav class="md-nav" aria-label="Dammit Commands">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#databases" class="md-nav__link">
    Databases
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    Annotation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-dammit" class="md-nav__link">
    Running Dammit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#modifying-params-for-dammit" class="md-nav__link">
    Modifying Params for Dammit:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-output" class="md-nav__link">
    Dammit Output
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-notes-non-elvers-parsing-dammit-gff3-files" class="md-nav__link">
    Additional Notes (non-elvers): Parsing Dammit GFF3 files
  </a>
  
</li>
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../sourmash/" title="sourmash" class="md-nav__link">
      sourmash
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../salmon/" title="salmon" class="md-nav__link">
      salmon
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../deseq2/" title="deseq2" class="md-nav__link">
      deseq2
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../plass/" title="plass" class="md-nav__link">
      plass
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../pear/" title="pear" class="md-nav__link">
      pear
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../paladin/" title="paladin" class="md-nav__link">
      paladin
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../rcorrector/" title="rcorrector" class="md-nav__link">
      rcorrector
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-nav__toggle md-toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6">
    
    <label class="md-nav__link" for="nav-6">
      For Developers
      <span class="md-nav__icon md-icon"></span>
    </label>
    <nav class="md-nav" aria-label="For Developers" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        <span class="md-nav__icon md-icon"></span>
        For Developers
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../under_the_hood/" title="Under the Hood" class="md-nav__link">
      Under the Hood
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#computational-requirements" class="md-nav__link">
    Computational Requirements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-commands" class="md-nav__link">
    Dammit Commands
  </a>
  
    <nav class="md-nav" aria-label="Dammit Commands">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#databases" class="md-nav__link">
    Databases
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    Annotation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-dammit" class="md-nav__link">
    Running Dammit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#modifying-params-for-dammit" class="md-nav__link">
    Modifying Params for Dammit:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-output" class="md-nav__link">
    Dammit Output
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-notes-non-elvers-parsing-dammit-gff3-files" class="md-nav__link">
    Additional Notes (non-elvers): Parsing Dammit GFF3 files
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="annotating-de-novo-transcriptomes-with-dammit">Annotating de novo transcriptomes with dammit<a class="headerlink" href="#annotating-de-novo-transcriptomes-with-dammit" title="Permanent link">&para;</a></h1>
<p><a href="http://dib-lab.github.io/dammit/">dammit</a> is an annotation pipeline written by <a href="http://www.camillescott.org/">Camille Scott</a>. dammit runs a relatively standard annotation protocol for transcriptomes: it begins by building gene models with <a href="http://transdecoder.github.io/">Transdecoder</a>, and then uses the following protein databases as evidence for annotation:</p>
<ul>
<li><a href="https://www.ebi.ac.uk/uniprot">Swiss-Prot</a> (manually reviewed and curated)</li>
<li><a href="http://pfam.xfam.org/">Pfam-A</a></li>
<li><a href="http://rfam.xfam.org/">Rfam</a></li>
<li><a href="http://www.orthodb.org/">OrthoDB</a></li>
<li><a href="http://www.uniprot.org/help/uniref">uniref90</a> (uniref is optional with <code>--full</code>).</li>
<li>nr (nr is optional with <code>--nr</code>).</li>
</ul>
<p>If a protein dataset is available, this can also be supplied to the
<code>dammit</code> pipeline with <code>--user-databases</code> as optional evidence for
annotation.</p>
<p>In addition, <a href="http://busco.ezlab.org/">BUSCO</a> v3 is run, which will compare the gene content in your transcriptome with a lineage-specific data set. The output is a proportion of your transcriptome that matches with the data set, which can be used as an estimate of the completeness of your transcriptome based on evolutionary expectation (<a href="http://bioinformatics.oxfordjournals.org/content/31/19/3210.full">Simho et al.2015</a>). There are several lineage-specific datasets available from the authors of BUSCO. We will use the <code>metazoa</code> dataset for this transcriptome.</p>
<h2 id="computational-requirements">Computational Requirements<a class="headerlink" href="#computational-requirements" title="Permanent link">&para;</a></h2>
<p>For the standard pipeline, dammit needs ~18GB of storage space to store its prepared databases, plus a few hundred MB per BUSCO database. For the standard annotation pipeline, we recommend at least 16GB of RAM. This can be reduced by editing LAST parameters via a custom configuration file.</p>
<p>The full pipeline, which uses uniref90, needs several hundred GB of space and considerable RAM to prepare the databases. You'll also want either a fat internet connection or a big cup of patience to download uniref.</p>
<p>For some species, we have found that the amount of RAM required can be proportional to the size of the transcriptome being annotated.</p>
<p>While dammit runs, it will print out which tasks its running to the terminal. dammit is written with a library called <a href="http://www.pydoit.org">pydoit</a>, which is a python workflow library similar to GNU Make and Snakemake. This not only helps organize the underlying workflow, but also means that if we interrupt it, it should properly resume! Caveat: if your job dies, without properly shutting down, snakemake will leave your directory "locked" (a safety feature to prevent two runs/programs from editing the same file simultaneously). If this happens, you'll need to run <code>elvers</code> with the <code>--unlock</code> flag.</p>
<h2 id="dammit-commands">Dammit Commands<a class="headerlink" href="#dammit-commands" title="Permanent link">&para;</a></h2>
<p>dammit has two major subcommands: <code>dammit databases</code> and <code>dammit annotate</code>. <code>databases</code> checks that the databases are installed and prepared, and if run with the <code>--install</code> flag, will perform that installation and preparation. <code>annotate</code> then performs the annotation using installed tools and databases. These commands are automated and integrated into a single rule in <code>elvers</code>. At some point, these may be split into <code>databases</code> and <code>annotate</code> if that functionality is desired, but it's not in
our immediate plans. </p>
<p>By default, databases are placed at <code>elvers/databases</code>.</p>
<p>Both <code>databases</code> and <code>annotate</code> have a <code>--quick</code> option, that only installs or runs a "quick" version of the pipeline: transdecoder</p>
<p>On the command line, the commands we would run are as follows:</p>
<h3 id="databases">Databases<a class="headerlink" href="#databases" title="Permanent link">&para;</a></h3>
<div class="codehilite"><pre><span></span><code>#install databases
dammit databases --install --busco-group metazoa --busco-group eukaryota
</code></pre></div>


<p>Note: if you have limited space on your instance, you can also install these databases in a different location (e.g. on an external volume) by adding <code>--database-dir /path/to/databases</code>. </p>
<p>You can also use a custom protein database for your species. If your critter is a non-model organism, you will likely need to create your own with proteins from closely-related species. This will rely on your knowledge of your system!</p>
<h3 id="annotation">Annotation<a class="headerlink" href="#annotation" title="Permanent link">&para;</a></h3>
<div class="codehilite"><pre><span></span><code>dammit annotate trinity.nema.fasta --busco-group metazoa --user-databases &lt;your-database-here&gt; --n_threads 6
</code></pre></div>


<p>If you want to run a quick version of the pipeline, add a parameter, <code>--quick</code>, to omit OrthoDB, Uniref, Pfam, and Rfam. A "full" run will take longer to install and run, but you'll have access to the full annotation pipeline.</p>
<h2 id="running-dammit">Running Dammit<a class="headerlink" href="#running-dammit" title="Permanent link">&para;</a></h2>
<p>Run Dammit via the "default" <a href="../eel_pond_workflow/">Eel Pond workflow</a> or via the <a href="../annotate/">annotate subworkflow</a>. To run Dammit as a standalone program, see "Advanced Usage" section below.</p>
<h2 id="modifying-params-for-dammit">Modifying Params for Dammit:<a class="headerlink" href="#modifying-params-for-dammit" title="Permanent link">&para;</a></h2>
<p>Be sure to set up your sample info and build a configfile first (see <a href="../configure/">Understanding and Configuring Workflows</a>).</p>
<p>To see the available parameters for the <code>dammit</code> rule, run</p>
<div class="codehilite"><pre><span></span><code>elvers config dammit --print_params
</code></pre></div>


<p>This will print the following:</p>
<div class="codehilite"><pre><span></span><code>  ####################  dammit  ####################
dammit:
  busco_group:     # specify all busco groups below here
  - metazoa
  - eukaryota
  db_dir: databases   # specify location for databases (or previously installed databases)
  db_install_only: False   # just install databases, don&#39;t run annotation
  db_extra: &#39;&#39;
  annot_extra: &#39; --quick &#39;
  #####################################################
</code></pre></div>


<p>In addition to changing parameters we've specifically enabled, you can modify the <code>extra</code> param to pass any extra parameters.</p>
<p>In dammit, both <code>databases</code> and <code>annotation</code> take an <code>extra</code> param:</p>
<div class="codehilite"><pre><span></span><code>  db_extra: &#39;--someflag someparam --someotherflag thatotherparam&#39;
  annot_extra: &#39;--someflag someparam --someotherflag thatotherparam&#39;
</code></pre></div>


<p>Within the "default" <a href="../eel_pond_workflow/">Eel Pond workflow</a>, this will annotate the Trinity assembly. Since <code>Dammit</code> requires an assembly as input, see the <a href="../annotate/">annotate subworkflow</a> instructions for how to run dammit as a standalone rule. See the <a href="http://dib-lab.github.io/dammit/">dammit documentation</a> for all params you can pass into Dammit.</p>
<h2 id="dammit-output">Dammit Output<a class="headerlink" href="#dammit-output" title="Permanent link">&para;</a></h2>
<p>Your main output directory will be determined by your config file: by default it is <code>BASENAME_out</code> (you specify BASENAME).</p>
<p>Dammit will output files in the <code>annotation</code> subdirectory of this output directory. The annotated fasta file will be <code>ASSEMBLY.dammit.fasta</code> and the annotation <code>gff3</code> file will be <code>ASSEMBLY.dammit.gff3</code>. Dammit will also produce a number of intermediate files that will be contained within an <code>ASSEMBLY.fasta.dammit</code> folder.</p>
<p>After a successful run, you'll have a new directory called <code>BASENAME.fasta.dammit</code>. If you look inside, you'll see a lot of files. For example, for a transcriptome with basename <code>trinity.nema</code>, the folder <code>trinity.nema.fasta.dammit</code> should contain the following files after a standard (not <code>--quick</code>) run:</p>
<div class="codehilite"><pre><span></span><code>ls trinity.nema.fasta.dammit/
</code></pre></div>


<div class="codehilite"><pre><span></span><code>    annotate.doit.db                              trinity.nema.fasta.dammit.namemap.csv  trinity.nema.fasta.transdecoder.pep
    dammit.log                                    trinity.nema.fasta.dammit.stats.json   trinity.nema.fasta.x.nema.reference.prot.faa.crbl.csv
    run_trinity.nema.fasta.metazoa.busco.results  trinity.nema.fasta.transdecoder.bed    trinity.nema.fasta.x.nema.reference.prot.faa.crbl.gff3
    tmp                                           trinity.nema.fasta.transdecoder.cds    trinity.nema.fasta.x.nema.reference.prot.faa.crbl.model.csv
    trinity.nema.fasta                            trinity.nema.fasta.transdecoder_dir    trinity.nema.fasta.x.nema.reference.prot.faa.crbl.model.plot.pdf
    trinity.nema.fasta.dammit.fasta               trinity.nema.fasta.transdecoder.gff3
    trinity.nema.fasta.dammit.gff3                trinity.nema.fasta.transdecoder.mRNA
</code></pre></div>


<p>As part of elvers, we copy the two most important files, <code>trinity.nema.fasta.dammit.fasta</code> and <code>trinity.nema.fasta.dammit.gff3</code> into the main <code>annotation</code> directory. <code>trinity.nema.fasta.dammit.stats.json</code> also gives summary stats that are quite useful.</p>
<p>If the above <code>dammit</code> command is run again, there will be a message:
<code>**Pipeline is already completed!**</code></p>
<p>If you'd like to rerun the dammit pipeline, you'll need to use the <code>--forceall</code> flag, like so:</p>
<div class="codehilite"><pre><span></span><code>elvers examples/nema.yaml annotation --forceall
</code></pre></div>


<h2 id="additional-notes-non-elvers-parsing-dammit-gff3-files">Additional Notes (non-elvers): Parsing Dammit GFF3 files<a class="headerlink" href="#additional-notes-non-elvers-parsing-dammit-gff3-files" title="Permanent link">&para;</a></h2>
<p>Camille wrote dammit in Python, which includes a library to parse gff3 dammit output. If you want to work with this gff3 downstream, use htis parsing library: </p>
<p>First, enter in a dammit environment. you can find the one elvers uses, but it might also be easier to just make a new one using the dammit environment file:</p>
<div class="codehilite"><pre><span></span><code># from within the elvers directory
conda env create -n dammit --file rules/dammit/dammit-env.yaml
source activate dammit
</code></pre></div>


<p>Remember you can exit your conda environments with <code>source deactivate</code></p>
<p>Then:</p>
<div class="codehilite"><pre><span></span><code>cd trinity.nema.fasta.dammit
python
</code></pre></div>


<p>Use gff3 librarys to output a list of gene IDs:</p>
<div class="codehilite"><pre><span></span><code>import pandas as pd
from dammit.fileio.gff3 import GFF3Parser
gff_file = &quot;trinity.nema.fasta.dammit.gff3&quot;
annotations = GFF3Parser(filename=gff_file).read()
names = annotations.sort_values(by=[&#39;seqid&#39;, &#39;score&#39;], ascending=True).query(&#39;score &lt; 1e-05&#39;).drop_duplicates(subset=&#39;seqid&#39;)[[&#39;seqid&#39;, &#39;Name&#39;]]
new_file = names.dropna(axis=0,how=&#39;all&#39;)
new_file.head()
new_file.to_csv(&quot;nema_gene_name_id.csv&quot;)
exit()
</code></pre></div>


<p>This will output a table of genes with 'seqid' and 'Name' in a .csv file: <code>nema_gene_name_id.csv</code>. Let's take a look at that file:</p>
<div class="codehilite"><pre><span></span><code>less nema_gene_name_id.csv
</code></pre></div>


<p>Notice there are multiple transcripts per gene model prediction. This <code>.csv</code> file can be used in <code>tximport</code> in downstream DE analysis.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../get_reference/" title="get_reference" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                get_reference
              </div>
            </div>
          </a>
        
        
          <a href="../sourmash/" title="sourmash" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                sourmash
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2018 <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a> at UC Davis <br/><br/> Icon made by Freepik, obtained from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a>, used under a <a href="http://creativecommons.org/licenses/by/3.0/" title="Creative Commons BY 3.0" target="_blank">CC 3.0 BY</a></div> license.
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/vendor.77e55a48.min.js"></script>
      <script src="../assets/javascripts/bundle.aa3f9871.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "..",
          features: ['instant'],
          search: Object.assign({
            worker: "../assets/javascripts/worker/search.4ac00218.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>