



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
        <link rel="canonical" href="http://dib-lab.github.io/eelpond/dammit/">
      
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching pages">
      
        <meta name="lang:search.result.one" content="1 matching page">
      
        <meta name="lang:search.result.other" content="# matching page">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="..">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-3.0.3">
    
    
      
        <title>dammit - eelpond</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.451f80e5.css">
      
        <link rel="stylesheet" href="../assets/stylesheets/application-palette.22915126.css">
      
      
        
        
        <meta name="theme-color" content="">
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../css/custom.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="darkblue" data-md-color-accent="blue">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="../#annotating-de-novo-transcriptomes-with-dammit" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="http://dib-lab.github.io/eelpond/" title="eelpond" class="md-header-nav__button md-logo">
          
            <i class="md-icon">dns</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                eelpond
              </span>
              <span class="md-header-nav__topic">
                dammit
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search this site" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching this site
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/dib-lab/eelpond" title="Go to repository" class="md-source" data-md-source="github">
    
    <div class="md-source__repository">
      eelpond
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="http://dib-lab.github.io/eelpond/" title="eelpond" class="md-nav__button md-logo">
      
        <i class="md-icon">dns</i>
      
    </a>
    eelpond
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/dib-lab/eelpond" title="Go to repository" class="md-source" data-md-source="github">
    
    <div class="md-source__repository">
      eelpond
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="About" class="md-nav__link">
      About
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      Available Workflows
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        Available Workflows
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../Eel_Pond_workflow/" title="Eel Pond Protocol" class="md-nav__link">
      Eel Pond Protocol
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../Protein_assembly_workflow/" title="PLASS Assembly" class="md-nav__link">
      PLASS Assembly
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2-3" type="checkbox" id="nav-2-3" checked>
    
    <label class="md-nav__link" for="nav-2-3">
      Programs Used
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="2">
      <label class="md-nav__title" for="nav-2-3">
        Programs Used
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../fastqc/" title="fastqc" class="md-nav__link">
      fastqc
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../trimmomatic/" title="trimmomatic" class="md-nav__link">
      trimmomatic
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../khmer/" title="khmer" class="md-nav__link">
      khmer
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../trinity/" title="trinity" class="md-nav__link">
      trinity
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        dammit
      </label>
    
    <a href="./" title="dammit" class="md-nav__link md-nav__link--active">
      dammit
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#quickstart-running-dammit-via-eelpond" title="Quickstart: Running dammit via eelpond" class="md-nav__link">
    Quickstart: Running dammit via eelpond
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit" title="dammit!" class="md-nav__link">
    dammit!
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#computational-requirements" title="Computational Requirements" class="md-nav__link">
    Computational Requirements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-commands" title="Dammit Commands" class="md-nav__link">
    Dammit Commands
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#databases" title="Databases" class="md-nav__link">
    Databases
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" title="Annotation" class="md-nav__link">
    Annotation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#customizing-dammit-parameters" title="Customizing Dammit parameters" class="md-nav__link">
    Customizing Dammit parameters
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-output" title="Dammit Output" class="md-nav__link">
    Dammit Output
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-notes-non-eelpond-parsing-dammit-gff3-files" title="Additional Notes (non-eelpond): Parsing Dammit GFF3 files" class="md-nav__link">
    Additional Notes (non-eelpond): Parsing Dammit GFF3 files
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../sourmash/" title="sourmash" class="md-nav__link">
      sourmash
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../salmon/" title="salmon" class="md-nav__link">
      salmon
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../deseq2/" title="deseq2" class="md-nav__link">
      deseq2
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../plass/" title="plass" class="md-nav__link">
      plass
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../pear/" title="pear" class="md-nav__link">
      pear
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../paladin/" title="paladin" class="md-nav__link">
      paladin
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../rcorrector/" title="rcorrector" class="md-nav__link">
      rcorrector
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3">
    
    <label class="md-nav__link" for="nav-3">
      For Developers
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        For Developers
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../dev_tips/" title="Notes for Developers" class="md-nav__link">
      Notes for Developers
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../Configuration/" title="Configuring the pipeline" class="md-nav__link">
      Configuring the pipeline
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#quickstart-running-dammit-via-eelpond" title="Quickstart: Running dammit via eelpond" class="md-nav__link">
    Quickstart: Running dammit via eelpond
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit" title="dammit!" class="md-nav__link">
    dammit!
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#computational-requirements" title="Computational Requirements" class="md-nav__link">
    Computational Requirements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-commands" title="Dammit Commands" class="md-nav__link">
    Dammit Commands
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#databases" title="Databases" class="md-nav__link">
    Databases
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" title="Annotation" class="md-nav__link">
    Annotation
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#customizing-dammit-parameters" title="Customizing Dammit parameters" class="md-nav__link">
    Customizing Dammit parameters
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dammit-output" title="Dammit Output" class="md-nav__link">
    Dammit Output
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-notes-non-eelpond-parsing-dammit-gff3-files" title="Additional Notes (non-eelpond): Parsing Dammit GFF3 files" class="md-nav__link">
    Additional Notes (non-eelpond): Parsing Dammit GFF3 files
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                
                <h1 id="annotating-de-novo-transcriptomes-with-dammit">Annotating de novo transcriptomes with dammit<a class="headerlink" href="#annotating-de-novo-transcriptomes-with-dammit" title="Permanent link">&para;</a></h1>
<h2 id="quickstart-running-dammit-via-eelpond">Quickstart: Running dammit via eelpond<a class="headerlink" href="#quickstart-running-dammit-via-eelpond" title="Permanent link">&para;</a></h2>
<p>Test data:</p>
<div class="codehilite"><pre><span></span>./run_eelpond nema-test annotate
</pre></div>


<p>Note that you either need to 1) have already run an assembly, such that a <code>fasta</code> file is sitting in the <code>eelpond/assembly</code> directory, 2) Run an assembly at the same time, or 3) pass an assembly in via <code>assemblyinput</code></p>
<p>If you have not already run <code>./run_eelpond nema-test assemble</code>:</p>
<p>2) Run trinity assembly at the same time:
   <code>./run_eelpond nema-test assemble annotate</code>
   3) OR, Pass an assembly in via <code>assemblyinput</code>
   <code>./run_eelpond assemblyinput annotate</code>
   with an assembly in your <code>yaml</code> configfile, e.g.:
   <code>assemblyinput:
     assembly: rna_testdata/nema.fasta
     gene_trans_map:  rna_testdata/nema.fasta.gene_trans_map
     assembly_extension: '_input'</code>
    This is commented out in the test data yaml, but go ahead and uncomment (remove leading <code>#</code>) in order to use this option</p>
<h2 id="dammit">dammit!<a class="headerlink" href="#dammit" title="Permanent link">&para;</a></h2>
<p><a href="http://www.camillescott.org/dammit/index.html">dammit</a> is an annotation pipeline written by <a href="http://www.camillescott.org/">Camille Scott</a>. dammit runs a relatively standard annotation protocol for transcriptomes: it begins by building gene models with <a href="http://transdecoder.github.io/">Transdecoder</a>, and then uses the following protein databases as evidence for annotation:
  -  <a href="https://www.ebi.ac.uk/uniprot">Swiss-Prot</a> (manually reviewed and curated)
  -  <a href="http://pfam.xfam.org/">Pfam-A</a>
  -  <a href="http://rfam.xfam.org/">Rfam</a>
  -  <a href="http://www.orthodb.org/">OrthoDB</a>
   - <a href="http://www.uniprot.org/help/uniref">uniref90</a> (uniref is optional with <code>--full</code>).
   - nr (nr is optional with <code>--nr</code>).</p>
<p>If a protein dataset is available, this can also be supplied to the
<code>dammit</code> pipeline with <code>--user-databases</code> as optional evidence for
annotation.</p>
<p>In addition, <a href="http://busco.ezlab.org/">BUSCO</a> v3 is run, which will compare the gene content in your transcriptome with a lineage-specific data set. The output is a proportion of your transcriptome that matches with the data set, which can be used as an estimate of the completeness of your transcriptome based on evolutionary expectation (<a href="http://bioinformatics.oxfordjournals.org/content/31/19/3210.full">Simho et al.2015</a>). There are several lineage-specific datasets available from the authors of BUSCO. We will use the <code>metazoa</code> dataset for this transcriptome.</p>
<h2 id="computational-requirements">Computational Requirements<a class="headerlink" href="#computational-requirements" title="Permanent link">&para;</a></h2>
<p>For the standard pipeline, dammit needs ~18GB of storage space to store its prepared databases, plus a few hundred MB per BUSCO database. For the standard annotation pipeline, we recommend at least 16GB of RAM. This can be reduced by editing LAST parameters via a custom configuration file.</p>
<p>The full pipeline, which uses uniref90, needs several hundred GB of space and considerable RAM to prepare the databases. You'll also want either a fat internet connection or a big cup of patience to download uniref.</p>
<p>For some species, we have found that the amount of RAM required can be proportional to the size of the transcriptome being annotated.</p>
<p>While dammit runs, it will print out which tasks its running to the terminal. dammit is written with a library called <a href="http://www.pydoit.org">pydoit</a>, which is a python workflow library similar to GNU Make and Snakemake. This not only helps organize the underlying workflow, but also means that if we interrupt it, it should properly resume! Caveat: if your job dies, without properly shutting down, snakemake will leave your directory "locked" (a safety feature to prevent two runs/programs from editing the same file simultaneously). If this happens, you'll need to run <code>eelpond</code> with the <code>--unlock</code> flag.</p>
<h2 id="dammit-commands">Dammit Commands<a class="headerlink" href="#dammit-commands" title="Permanent link">&para;</a></h2>
<p>dammit has two major subcommands: <code>dammit databases</code> and <code>dammit annotate</code>. <code>databases</code> checks that the databases are installed and prepared, and if run with the <code>--install</code> flag, will perform that installation and preparation. <code>annotate</code> then performs the annotation using installed tools and databases. These commands are automated and integrated into a single rule in <code>eelpond</code>. At some point, these may be split into <code>databases</code> and <code>annotate</code> if that functionality is desired, but it's not in
our immediate plans. </p>
<p>By default, databases are placed at <code>eelpond/databases</code>.</p>
<p>Both <code>databases</code> and <code>annotate</code> have a <code>--quick</code> option, that only installs or runs a "quick" version of the pipeline: transdecoder</p>
<p>On the command line, the commands we would run are as follows:</p>
<h3 id="databases">Databases<a class="headerlink" href="#databases" title="Permanent link">&para;</a></h3>
<div class="codehilite"><pre><span></span>#install databases
dammit databases --install --busco-group metazoa --busco-group eukaryota
</pre></div>


<p>Note: if you have limited space on your instance, you can also install these databases in a different location (e.g. on an external volume) by adding <code>--database-dir /path/to/databases</code>. </p>
<p>You can also use a custom protein database for your species. If your critter is a non-model organism, you will likely need to create your own with proteins from closely-related species. This will rely on your knowledge of your system!</p>
<h3 id="annotation">Annotation<a class="headerlink" href="#annotation" title="Permanent link">&para;</a></h3>
<div class="codehilite"><pre><span></span>dammit annotate trinity.nema.fasta --busco-group metazoa --user-databases &lt;your-database-here&gt; --n_threads 6
</pre></div>


<p>If you want to run a quick version of the pipeline, add a parameter, <code>--quick</code>, to omit OrthoDB, Uniref, Pfam, and Rfam. A "full" run will take longer to install and run, but you'll have access to the full annotation pipeline.</p>
<h2 id="customizing-dammit-parameters">Customizing Dammit parameters<a class="headerlink" href="#customizing-dammit-parameters" title="Permanent link">&para;</a></h2>
<p>To modify any program params, you need to add a couple lines to the config file you provide to <code>eelpond</code>.</p>
<p>To get a dammit configfile you can modify, run:</p>
<div class="codehilite"><pre><span></span>./run_eelpond dammit.yaml dammit --build_config
</pre></div>


<p>The output should be a small <code>yaml</code> configfile that will look something like this:</p>
<div class="codehilite"><pre><span></span>  ####################  dammit  ####################
dammit:
  busco_group:
  - metazoa
  - eukaryota
  db_dir: databases
  db_extra:
  annot_extra: &#39; --quick &#39;
</pre></div>


<p>In addition to changing parameters we've specifically enabled, you can modify the <code>extra</code> param to pass any extra parameters.</p>
<p>In dammit, both <code>databases</code> and <code>annotation</code> take an <code>extra</code> param:</p>
<div class="codehilite"><pre><span></span>  db_extra: &#39;--someflag someparam --someotherflag thatotherparam&#39;
  annot_extra: &#39;--someflag someparam --someotherflag thatotherparam&#39;
</pre></div>


<p>Override default params by modifying any of these lines, and placing them in the config file you're using to run <code>eelpond</code>.</p>
<h2 id="dammit-output">Dammit Output<a class="headerlink" href="#dammit-output" title="Permanent link">&para;</a></h2>
<p>After a successful run, you'll have a new directory called <code>BASENAME.fasta.dammit</code>. If you look inside, you'll see a lot of files. For example, for a transcriptome with basename <code>trinity.nema</code>, the folder <code>trinity.nema.fasta.dammit</code> should contain the following files after a standard (not <code>--quick</code>) run:</p>
<div class="codehilite"><pre><span></span>ls trinity.nema.fasta.dammit/
</pre></div>


<div class="codehilite"><pre><span></span>    annotate.doit.db                              trinity.nema.fasta.dammit.namemap.csv  trinity.nema.fasta.transdecoder.pep
    dammit.log                                    trinity.nema.fasta.dammit.stats.json   trinity.nema.fasta.x.nema.reference.prot.faa.crbl.csv
    run_trinity.nema.fasta.metazoa.busco.results  trinity.nema.fasta.transdecoder.bed    trinity.nema.fasta.x.nema.reference.prot.faa.crbl.gff3
    tmp                                           trinity.nema.fasta.transdecoder.cds    trinity.nema.fasta.x.nema.reference.prot.faa.crbl.model.csv
    trinity.nema.fasta                            trinity.nema.fasta.transdecoder_dir    trinity.nema.fasta.x.nema.reference.prot.faa.crbl.model.plot.pdf
    trinity.nema.fasta.dammit.fasta               trinity.nema.fasta.transdecoder.gff3
    trinity.nema.fasta.dammit.gff3                trinity.nema.fasta.transdecoder.mRNA
</pre></div>


<p>As part of eelpond, we copy the two most important files, <code>trinity.nema.fasta.dammit.fasta</code> and <code>trinity.nema.fasta.dammit.gff3</code> into the main <code>annotation</code> directory. <code>trinity.nema.fasta.dammit.stats.json</code> also gives summary stats that are quite useful.</p>
<p>If the above <code>dammit</code> command is run again, there will be a message:
<code>**Pipeline is already completed!**</code></p>
<p>If you'd like to rerun the dammit pipeline, you'll need to use the <code>--forceall</code> flag, like so:</p>
<div class="codehilite"><pre><span></span>./run_eelpond nema-test annotation --forceall
</pre></div>


<h2 id="additional-notes-non-eelpond-parsing-dammit-gff3-files">Additional Notes (non-eelpond): Parsing Dammit GFF3 files<a class="headerlink" href="#additional-notes-non-eelpond-parsing-dammit-gff3-files" title="Permanent link">&para;</a></h2>
<p>Camille wrote dammit in Python, which includes a library to parse gff3 dammit output. If you want to work with this gff3 downstream, use htis parsing library: </p>
<p>First, enter in a dammit environment. you can find the one eelpond uses, but it might also be easier to just make a new one using the dammit environment file:</p>
<div class="codehilite"><pre><span></span># from within the eelpond directory
conda env create -n dammit --file rules/dammit/dammit-env.yaml
source activate dammit
</pre></div>


<p>Remember you can exit your conda environments with <code>source deactivate</code></p>
<p>Then:</p>
<div class="codehilite"><pre><span></span>cd trinity.nema.fasta.dammit
python
</pre></div>


<p>Use gff3 librarys to output a list of gene IDs:</p>
<div class="codehilite"><pre><span></span>import pandas as pd
from dammit.fileio.gff3 import GFF3Parser
gff_file = &quot;trinity.nema.fasta.dammit.gff3&quot;
annotations = GFF3Parser(filename=gff_file).read()
names = annotations.sort_values(by=[&#39;seqid&#39;, &#39;score&#39;], ascending=True).query(&#39;score &lt; 1e-05&#39;).drop_duplicates(subset=&#39;seqid&#39;)[[&#39;seqid&#39;, &#39;Name&#39;]]
new_file = names.dropna(axis=0,how=&#39;all&#39;)
new_file.head()
new_file.to_csv(&quot;nema_gene_name_id.csv&quot;)
exit()
</pre></div>


<p>This will output a table of genes with 'seqid' and 'Name' in a .csv file: <code>nema_gene_name_id.csv</code>. Let's take a look at that file:</p>
<div class="codehilite"><pre><span></span>less nema_gene_name_id.csv
</pre></div>


<p>Notice there are multiple transcripts per gene model prediction. This <code>.csv</code> file can be used in <code>tximport</code> in downstream DE analysis.</p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../trinity/" title="trinity" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                trinity
              </span>
            </div>
          </a>
        
        
          <a href="../sourmash/" title="sourmash" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                sourmash
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright &copy; 2018 <a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a> at UC Davis
          </div>
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.e72fd936.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
        <script src="../search/main.js"></script>
      
    
    
      
    
  </body>
</html>