# FastQC


## Quickstart: Running FastQC with eelpond
```
./run_eelpond nema-test preprocess
```
This will run trimmomatic trimming and fastqc on pre-trim and post-trim data.

## FastQC

We use [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) to assess quality of sequencing data before and after adapter trimming. 

From the [FastQC documentation](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/):  
```
What is FastQC
Modern high throughput sequencers can generate hundreds of millions of sequences in a single run. Before analysing this sequence to draw biological conclusions you should always perform some simple quality control checks to ensure that the raw data looks good and there are no problems or biases in your data which may affect how you can usefully use it.

Most sequencers will generate a QC report as part of their analysis pipeline, but this is usually only focused on identifying problems which were generated by the sequencer itself. FastQC aims to provide a QC report which can spot problems which originate either in the sequencer or in the starting library material.

FastQC can be run in one of two modes. It can either run as a stand alone interactive application for the immediate analysis of small numbers of FastQ files, or it can be run in a non-interactive mode where it would be suitable for integrating into a larger analysis pipeline for the systematic processing of large numbers of files.
```

There are several caveats about FastQC - the main one is that it only calculates certain statistics (like duplicated sequences) for subsets of the data (e.g. duplicate sequences are only analyzed for the first 100,000 sequences in each file). 

Check out these examples of [good](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html) and [bad](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html) Illumina data.


## FastQC Params

To modify any program params, you need to add a couple lines to the config file you provide to `eelpond`. 

To get a fastqc config you can modify, run:
```
./run_eelpond fastqc.yaml fastqc --build_config
```
The output should be a small `yaml` configfile that contains:
```
  ####################  fastqc  ####################
fastqc:
  extra: ''
```
There's almost nothing in here because we use default params. However, you can modify the `extra` param to pass any extra trimmomatic parameters, e.g.:
```
  extra: '--someflag someparam --someotherflag thatotherparam'
```
Be sure the modified lines go into the config file you're using to run `eelpond`.

## FastQC rule

We use a local copies of the [fastqc snakemake wrapper](https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/fastqc.html) to run Fastqc.

For snakemake afficionados, Here's the basic structure of the fastqc rules. Directories and parameters are specified via the configfile (see the rule on [github](https://github.com/dib-lab/eelpond/blob/master/rules/fastqc/fastqc.rule)).


```
def get_trimmed(wildcards):
    if not is_single_end(**wildcards):
        return expand(join(trim_dir, '{sample}_{unit}_{end}.fq.gz'), **wildcards)
    return expand(join(trim_dir, '{sample}_{unit}_{end}.fq.gz'), **wildcards)

def get_pretrim(wildcards):
    if not is_single_end(**wildcards):
        return expand(join(data_dir, '{sample}_{unit}_{end}.fq.gz'), **wildcards)
    return expand(join(data_dir, '{sample}_{unit}_{end}.fq.gz'), **wildcards)

rule fastqc_trimmed:
    input:
       get_trimmed
    output:
        html=join(qc_dir,'{sample}_{unit}_{end}.fastqc.html'),
        zip=join(qc_dir,'{sample}_{unit}_{end}.fastqc.zip')
    params: fastqc_params.get('extra', "")
    log:
        join(LOGS_DIR, 'fastqc/{sample}_{unit}_{end}.log')
    conda: "fastqc-env.yaml"
    script: "wrapper.py"

rule fastqc_pretrim:
    input:
        get_pretrim
    output:
        html=join(qc_dir,'{sample}_{unit}_{end}.fastqc.html'),
        zip=join(qc_dir,'{sample}_{unit}_{end}.fastqc.zip')
    params: fastqc_params.get('extra', "")
    log:
        join(LOGS_DIR, 'fastqc/{sample}_{unit}_{end}.log')
    conda: "fastqc-env.yaml"
    script: "wrapper.py"
```

